_source/_posts/2022-03-03-spring-native-jhipster.adoc

```diff
@@ -0,0 +1,500 @@
+        log.debug("REST request to save Photo : {}", photo);
+        if (photo.getId() != null) { ... }
+
+        try {
+            photo = setMetadata(photo);
+        } catch (ImageProcessingException | IOException | MetadataException ipe) {
+            log.error(ipe.getMessage());
+        }
+
+        Photo result = photoRepository.save(photo);
+        ...
+    }
+
+    private Photo setMetadata(Photo photo) throws ImageProcessingException, IOException, MetadataException {
+        String str = DatatypeConverter.printBase64Binary(photo.getImage());
+        byte[] data2 = DatatypeConverter.parseBase64Binary(str);
+        InputStream inputStream = new ByteArrayInputStream(data2);
+        BufferedInputStream bis = new BufferedInputStream(inputStream);
+        Metadata metadata = ImageMetadataReader.readMetadata(bis);
+        ExifSubIFDDirectory directory = metadata.getFirstDirectoryOfType(ExifSubIFDDirectory.class);
+
+        if (directory != null) {
+            Date date = directory.getDateDigitized();
+            if (date != null) {
+                photo.setTaken(date.toInstant());
+            }
+        }
+
+        if (photo.getTaken() == null) {
+            log.debug("Photo EXIF date digitized not available, setting taken on date to now...");
+            photo.setTaken(Instant.now());
+        }
+
+        photo.setUploaded(Instant.now());
+
+        JpegDirectory jpgDirectory = metadata.getFirstDirectoryOfType(JpegDirectory.class);
+        if (jpgDirectory != null) {
+            photo.setHeight(jpgDirectory.getImageHeight());
+            photo.setWidth(jpgDirectory.getImageWidth());
+        }
+
+        return photo;
+    }
+    ...
+}
+----
+
+. Install the React libraries needed:
++
+[source,shell]
+----
+npm i react-photo-gallery@8 --force
+npm i react-images
+----
+
+. In `src/test/javascript/cypress/integration/entity/photo.spec.ts`, remove the code that sets the calculated data in the `should create an instance of Photo` test:
++
+[source,typrescript]
+----
+cy.get(`[data-cy="height"]`).type('99459').should('have.value', '99459');
+cy.get(`[data-cy="width"]`).type('61514').should('have.value', '61514');
+cy.get(`[data-cy="taken"]`).type('2021-10-11T16:46').should('have.value', '2021-10-11T16:46');
+cy.get(`[data-cy="uploaded"]`).type('2021-10-11T15:23').should('have.value', '2021-10-11T15:23');
+----
+
+Then, you'll need to add type hints for Drew Noake's EXIF processing library in `src/main/java/com/auth0/flickr2/Flickr2App.java`:
+
+[source,java]
+----
+@org.springframework.nativex.hint.TypeHint(
+types = {
+    ...
+    com.drew.metadata.exif.ExifIFD0Directory.class,
+    com.drew.metadata.exif.ExifSubIFDDirectory.class,
+    com.drew.metadata.exif.ExifThumbnailDirectory.class,
+    com.drew.metadata.exif.makernotes.AppleMakernoteDirectory.class,
+    com.drew.metadata.exif.GpsDirectory.class,
+})
+@org.springframework.nativex.hint.NativeHint(options = "-H:+AddAllCharsets")
```

Comment:
```suggestion
@org.springframework.nativex.hint.TypeHint(
    types = {
        ...
        com.drew.metadata.exif.ExifIFD0Directory.class,
        com.drew.metadata.exif.ExifSubIFDDirectory.class,
        com.drew.metadata.exif.ExifThumbnailDirectory.class,
        com.drew.metadata.exif.makernotes.AppleMakernoteDirectory.class,
        com.drew.metadata.exif.GpsDirectory.class,
})
@org.springframework.nativex.hint.NativeHint(options = "-H:+AddAllCharsets")
```

`types` _should_ be indented, but take that with a grain of 🧂 and ignore this if you want 😄